src/charm.py

```diff
@@ -516,6 +522,81 @@ def _on_list_backups_action(self, event: ActionEvent) -> None:
 
         event.set_results({"backup-ids": json.dumps(backup_ids)})
 
+    def _on_restore_backup_action(self, event: ActionEvent) -> None:  # noqa: C901
```

Comment:
Could we split this a bit? Ignoring `C901` is a big smell that things can be simplified. I think we could this method describe the process at a higher level:
```
if not self._check_prerequisites(event):
   return
if not (snapshot := self._get_snapshot_from_s3()):
   return
if not self._restore_snapshot(snapshot):
   event.fail()
event.success()
```

I think this would make this easier to read, as right now we have many different levels of abstraction in the same method and a lot of context to keep in mind when reading it.

Reply:
Do you recommend the action to be able to fail in multiple places? It feels somewhat strange to move the whole event object around. I rather have the responsibility of success/failure be centralized in one function. It is not something we do anywhere else in the code in either the machine or K8s Vault charms, and for what it's worth we also ignore the C901 in the K8s charm. In any case, I would suggest not  starting this pattern here.

This being said, we could have a  `self._check_s3_prerequisites` that returns an error string (or something like it). It feels quite go-y but it would prevent the C901. For example:
```
def _on_restore_backup_action(self, event):
    err = self._check_s3_pre_requisites():
    if err != "":
        event.fail(message="s3 pre-requisites are not met: {error}")
        return
    ...
```

What do you think?

Reply:
Yes, I think keeping the failure/success of the event in this method a good approach. I agree that passing the event around is not the best way forward, and if any of the methods need something from the event, we can simply pass that exact thing instead of the whole event.

Reply:
Done, we don't have that C901 anymore